pkg/chunk/purger/delete_requests_store.go

```diff
@@ -212,8 +212,9 @@ func (ds *DeleteStore) GetPendingDeleteRequestsForUser(ctx context.Context, user
 
 func (ds *DeleteStore) queryDeleteRequests(ctx context.Context, deleteQuery []chunk.IndexQuery) ([]DeleteRequest, error) {
 	deleteRequests := []DeleteRequest{}
+	var itr chunk.ReadBatchIterator
 	err := ds.indexClient.QueryPages(ctx, deleteQuery, func(query chunk.IndexQuery, batch chunk.ReadBatch) (shouldContinue bool) {
-		itr := batch.Iterator()
+		itr = batch.Iterator(itr)
```

Comment:
I think this is introducing a bug, if there are multiple `delete queries`, since this iterator will now be shared between multiple concurrently running goroutines.

If there is only one, there is no need to reuse iterator.

Access to `deleteRequests` is not protected either. :confused:

Reply:
I sent a PR to rework this function to accept only one request.